### PR TITLE
SCMOD-14011: Support file-based secrets

### PR DESCRIPTION
**Pull request by rorytorneymf**
https://portal.digitalsafe.net/browse/SCMOD-14011

### CODE REVIEW

File: src/main/docker/startup/startup.sh [hunk 1]
```diff
@@ -20,6 +20,9 @@ log() {
     echo "[$(date +%H:%M:%S.%3NZ) #$(printf '%03X\n' $$).??? INFO  -            -   ] ${0##*/}: $@"
 }
 
+# Export file based secrets
+exec $(dirname "$0")/../scripts/export-file-based-secrets.sh
```
Comment (dermot-hardy) [inline src/main/docker/startup/startup.sh]: Does this actually work?  I would have thought that using `exec` would cause the rest of the script to not be executed?

File: src/main/docker/scripts/export-file-based-secrets.sh [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+
+# Create a convenience function for info logging.
+info() {
+    echo "[$(date +%H:%M:%S.%3NZ) #$(printf '%03X\n' $$).??? INFO  -            -   ] ${0##*/}: $@"
+}
+
+# Create a convenience function for error logging.
+error() {
+    echo "[$(date +%H:%M:%S.%3NZ) #$(printf '%03X\n' $$).??? ERROR  -            -   ] ${0##*/}: $@"
+}
```
Comment (dermot-hardy) [inline src/main/docker/scripts/export-file-based-secrets.sh]: Just use `echo` instead of these functions, and then when calling the script pipe it through the `caf-log-format.sh` script.  That way we'll keep the logging format encapsulated.

File: src/main/docker/scripts/export-file-based-secrets.sh [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+#
+#     ABC_PASSWORD_FILE=/var/somefile.txt
+#
+# read the contents of /var/somefile.txt (for example 'mypassword'), and export an environment variable named ABC_PASSWORD set to:
+#
+#     ABC_PASSWORD=mypassword
+export_file_based_secrets() {
+    env | while IFS= read -r env_var; do
+        local env_var_name=${env_var%%=*}
+        local env_var_value=${env_var#*=}
```
Comment (dermot-hardy) [inline src/main/docker/scripts/export-file-based-secrets.sh]: This would be easily spoofed (by having an environment variable containing a newline).  You should instead use `env -0`:

```suggestion
    env -0 | while IFS='=' read -r -d '' env_var_name env_var_value; do
```

File: src/main/docker/scripts/export-file-based-secrets.sh [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+                echo "ERROR: Both $env_var_name and $env_var_name_without_file_suffix are set (but are exclusive)"
+                exit 1
+            fi            
+            echo "INFO: Found environment variable ending with the _FILE suffix: $env_var_name=$env_var_value, attempting to read the \
+contents of $env_var_value..."
+            if [ -e "$env_var_value" ]; then
+                local file_contents=$(<${env_var_value})
+                echo "INFO: Successfully read contents of ${env_var_value}, exporting environment variable \
+$env_var_name_without_file_suffix using the contents of ${env_var_value} as the value..."
+                if export "$env_var_name_without_file_suffix"="$file_contents" ; then
```
Comment (rorytorneymf) [inline src/main/docker/scripts/export-file-based-secrets.sh:44]: I'm not sure the `export` is working. When I created a derived image:

Dockerfile
```
FROM dev/cafapi/opensuse-base:2.4.0-SNAPSHOT
```

Built it:
```
docker build -t test .
```

And the ran it passing it the `env` command:
```
docker run -e ABC_PASSWORD_FILE=/pw.txt -d test env
[14:40:54.825Z #006.??? INFO  -            -   ] startup.sh: Running export-file-based-secrets.sh...
[14:40:54.831Z #00A.??? INFO  -            -   ] export-file-based-secrets.sh: Found environment variable ending with the _FILE suffix: ABC_PASSWORD_FILE=/pw.txt, attempting to read the contents of /pw.txt...
[14:40:54.834Z #00A.??? INFO  -            -   ] export-file-based-secrets.sh: Successfully read contents of /pw.txt, exporting environment variable ABC_PASSWORD using the contents of /pw.txt as the value...
[14:40:54.837Z #00A.??? INFO  -            -   ] export-file-based-secrets.sh: Successfully exported environment variable ABC_PASSWORD
...
LANG=en_US.utf8
HOSTNAME=403449a86633
TINI_VERSION=v0.18.0
ABC_PASSWORD_FILE=/pw.txt
PWD=/
HOME=/root
SHLVL=0
PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
```

the `ABC_PASSWORD` env var is not there (just the `ABC_PASSWORD_FILE` env var passed to the run command)

I'll need to look into it a bit more.

Comment (rorytorneymf) [reply src/main/docker/scripts/export-file-based-secrets.sh:44]: Should be good now, had to rework a couple of things: https://github.com/CAFapi/opensuse-base-image/pull/34/commits/15a66024b22b4df3bc9e434e0bc34e6f80215da5

File: src/main/docker/scripts/export-file-based-secrets.sh [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+# A function for logging in the caf logging format.
+log() {
+    echo "$@" |& $(dirname "$0")/../scripts/caf-log-format.sh "export-file-based-secrets.sh"
+}
+
```
Comment (dermot-hardy) [inline src/main/docker/scripts/export-file-based-secrets.sh]: Just directly use `echo` instead.  When calling the script pipe it through the `caf-log-format.sh` script and it will handle putting it into the right format.

**Review by dermot-hardy — CHANGES_REQUESTED** [review]

File: src/main/docker/startup/startup.sh [hunk 6]
```diff
@@ -20,6 +20,16 @@ log() {
...
 
+# Export file based secrets
+log "Running export-file-based-secrets.sh..."
+source $(dirname "$0")/../scripts/export-file-based-secrets.sh
+export_file_based_secrets_status=${PIPESTATUS[0]}
+if [ $export_file_based_secrets_status -ne 0 ]; then
+    echo "ERROR: Error running export-file-based-secrets.sh" |& $(dirname "$0")/../scripts/caf-log-format.sh "startup.sh"
+    exit $export_file_based_secrets_status
+fi
+log "export-file-based-secrets.sh completed"
```
Comment (dermot-hardy) [inline src/main/docker/startup/startup.sh]: Drop this.  `Running startup scripts...` covers it.

File: src/main/docker/scripts/export-file-based-secrets.sh (hunk 5 above)
Comment (rorytorneymf) [reply src/main/docker/scripts/export-file-based-secrets.sh]: I couldn't find a way to make that work, for example, this works:

```
source $(dirname "$0")/../scripts/export-file-based-secrets.sh
```

But piping this script to `caf-log-format.sh` doesn't work - I believe because when you pipe to a command, a subshell is created and any variables set inside of that subshell will go away when it exits. When I do this, the variables are not exported:

```
source $(dirname "$0")/../scripts/export-file-based-secrets.sh |& $(dirname "$0")/../scripts/caf-log-format.sh "export-file-based-secrets.sh"
```

Comment (dermot-hardy) [reply src/main/docker/scripts/export-file-based-secrets.sh]: Ok fair enough.

File: README.md [hunk 7]
```diff
@@ -31,6 +31,20 @@ Any executable scripts added to the `/startup/startup.d/` directory will be auto
 #### Certificate Installation
 The image comes pre-installed with a startup script which provides a mechanism to extend the CA certificates which should be trusted.
 
+#### Export File-Based Secrets Script
```
Comment (rorytorneymf) [inline README.md:34]: @michael-bryson / @dermot-hardy Should I update the readme + release notes of the other base images (https://github.com/CAFapi/opensuse-python3-image etc) with this info as well?

File: src/main/docker/scripts/export-file-based-secrets.sh [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+# A function for logging in the caf logging format.
+log() {
+    echo "$@" |& $(dirname "$0")/../scripts/caf-log-format.sh "export-file-based-secrets.sh"
```
Comment (dermot-hardy) [inline src/main/docker/scripts/export-file-based-secrets.sh:20]:
```suggestion
    echo "$@" |& $(dirname "$0")/../scripts/caf-log-format.sh "${0##*/}"
```

File: src/main/docker/scripts/export-file-based-secrets.sh [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+            local env_var_name_without_file_suffix=${env_var_name%_FILE}
+            if [ "${!env_var_name:-}" ] && [ "${!env_var_name_without_file_suffix:-}" ]; then
+                log "ERROR: Both $env_var_name and $env_var_name_without_file_suffix are set (but are exclusive)"
+                exit 1
+            fi            
+            log "INFO: Found environment variable ending with the _FILE suffix: $env_var_name=$env_var_value, attempting to read the \
+contents of $env_var_value..."
+            if [ -e "$env_var_value" ]; then
+                local file_contents=$(<${env_var_value})
+                log "INFO: Successfully read contents of ${env_var_value}, exporting environment variable \
```
Comment (dermot-hardy) [inline src/main/docker/scripts/export-file-based-secrets.sh]: Is this log statement really needed?  I feel like there is too much logging here.  You're going to log this just below anyway.

File: src/main/docker/scripts/export-file-based-secrets.sh [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+                if export "$env_var_name_without_file_suffix"="$file_contents" ; then
+                    log "INFO: Successfully exported environment variable $env_var_name_without_file_suffix"
+                    unset "$env_var_name"
+                else
+                    log "ERROR: Failed to export environment variable $env_var_name_without_file_suffix"
+                    exit 1
+                fi
+            else 
+                log "ERROR: Failed to export the environment variable $env_var_name_without_file_suffix, file $env_var_value does not \
+exist"
```
Comment (dermot-hardy) [inline src/main/docker/scripts/export-file-based-secrets.sh]: Shorten this to fit on one line.

**Review by dermot-hardy — APPROVED** [review]

File: README.md (hunk 7 above)
Comment (michael-bryson) [reply README.md:34]: Yes that's what we have done before as users will consume the base image they want so its better to have details of all the scripts in each base image readme.

Comment (dermot-hardy) [reply README.md:34]: If they have an inherited section on `Certificate Installation` then I guess so.

File: src/main/docker/scripts/export-file-based-secrets.sh (hunk 8 above)
Comment (rorytorneymf) [reply src/main/docker/scripts/export-file-based-secrets.sh:20]: Because the `export-file-based-secrets.sh` script is sourced by `startup.sh`, this change was causing the logs echoed by `export-file-based-secrets.sh` to report that they were coming from `startup.sh`. 

I also found another bug in that I am sourcing this file (and thus this `log` function) which was replacing the `log` function in `startup.sh`.

I'm not sure what the best thing to do here is? I was thinking:

1) Rename the `log` function in this script to `caf_log`
2) Hardcode the `export-file-based-secrets.sh` name in the function as I had previously done

```
caf_log() {
    echo "$@" |& $(dirname "$0")/../scripts/caf-log-format.sh "export-file-based-secrets.sh"
}
```
3) As an extra measure, unset the `caf_log` function at the end of  `export-file-based-secrets.sh`, so that it's not sourced into `startup.sh`

```
unset -f caf_log # Don't export the caf_log function when this script is sourced
```

This way the script names are correct in the logs:

```
[14:36:34.348Z #006.??? INFO  -            -   ] startup.sh: Running export-file-based-secrets.sh...
[14:36:34.357Z #00D.??? INFO  -            -   ] export-file-based-secrets.sh: Found environment variable ending with the _FILE suffix: ABC_PASSWORD_FILE=pw.txt, attempting to read the contents of pw.txt...
[14:36:34.364Z #016.??? INFO  -            -   ] export-file-based-secrets.sh: Successfully exported environment variable ABC_PASSWORD
[14:36:34.366Z #006.??? INFO  -            -   ] startup.sh: Running startup scripts...
```

Comment (rorytorneymf) [reply src/main/docker/scripts/export-file-based-secrets.sh:20]: Tried this out here: https://github.com/CAFapi/opensuse-base-image/pull/34/commits/ce7a858d25df96680fda894eb67ca702a3a0e5b1 

Happy to change it though if you think there's a better way.

File: src/main/docker/scripts/export-file-based-secrets.sh [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+    while IFS='=' read -r -d '' env_var_name env_var_value; do
+        if  [[ ${env_var_name} == *_FILE ]] ;
+        then
+            local env_var_name_without_file_suffix=${env_var_name%_FILE}
+            if [ "${!env_var_name:-}" ] && [ "${!env_var_name_without_file_suffix:-}" ]; then
+                caf_log "ERROR: Both $env_var_name and $env_var_name_without_file_suffix are set (but are exclusive)"
+                exit 1
+            fi            
+            caf_log "INFO: Found environment variable ending with the _FILE suffix: $env_var_name=$env_var_value, attempting to read \
+contents of $env_var_value..."
```
Comment (dermot-hardy) [inline src/main/docker/scripts/export-file-based-secrets.sh]: Shorten this:
```suggestion
            caf_log "INFO: Reading ${env_var_name} (${env_var_value})..."
```

**Review by dermot-hardy — APPROVED** [review]

File: src/main/docker/scripts/export-file-based-secrets.sh (hunk 8 above)
Comment (dermot-hardy) [reply src/main/docker/scripts/export-file-based-secrets.sh:20]: It's fine.  There are a lot more subtleties here than I realized.
